File: 046. Permutations/Java/Solution.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        result.add(new ArrayList<>());
        for (int i = 0; i < nums.length; i++) {
            List<List<Integer>> newResult = new ArrayList<>();
            for (int j = 0; j <= i; j++) {
                for (List<Integer> item : result) {
                    List<Integer> newItem = new ArrayList<>(item);
                    newItem.add(j, nums[i]);
                    newResult.add(newItem);
                }
            }
            result = newResult;
        }
        return result;
    }
}

File: 046. Permutations/C++/Solution.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int> &nums) {
        vector<vector<int>> result;
        result.push_back(vector<int>());
        for (int i = 0; i < nums.size(); i++) {
            vector<vector<int>> newResult;
            for (int j = 0; j <= i; j++) {
                for (auto item : result) {
                    vector<int> newItem(item);
                    newItem.insert(newItem.begin() + j, nums[i]);
                    newResult.push_back(newItem);
                }
            }
            result = newResult;
        }
        return result;
    }
};